### ISSUE TITLE
Running setup.py --help or clean will build cython files

### ISSUE THREAD

**Issue by infinity0** [cython/cython#1495]
(Originally reported as fplll/fpylll#46)

The [recommended way](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#multiple-cython-files-in-a-package) to build cython files will run the build even if you are calling `setup.py clean` or `setup.py --help`. This can waste quite a bit of development time. It would be better if the cythonizing process is only run during an actual build.

**Comment by infinity0** [comment]
Also, `clean` does not actually clean the cython-generated files.

**Comment by bjodah** [comment]
You can manually inspect arguments passed to `setup.py` and only run `cythonize` if it makes sense.
Here is an example: https://github.com/bjodah/pyodeint/blob/98b806aa0a330e0cea91be950de0b110fa64a5ed/setup.py#L25

**Comment by infinity0** [comment]
Right, but it should be possible to do this in a simpler way, that is simple enough to put in the official documentation. People aren't going to Do The Right Thing if it takes them too much time. Also, I think your example still doesn't actually clean the cython-built files, since setuptools isn't aware of that. The proper way to fix this would be to do hook into the extension points that setuptools provides. I forget what they are exactly, but they do have documentation on this.

**Comment by bjodah** [comment]
You are right, it does nothing about clean up. And I agree that it would be nice if this was done auto-magically (by cython/setuptools) for the `setup.py` author.

**Comment by robertwb** [comment]
There's not really a "good" way to do this. I would use `Cython.Build.Distutils.build_ext` to cythonize on build, or ship the C files. 

What we need to do is get (some of) this logic into distutils/setuptools.

**Comment by jdemeyer** [comment]
CC @embray

**Comment by embray** [comment]
I have bee meaning to raise this issue as well, as the docs Cython provides, which demonstrates running `cythonize` directly in the body of the `setup.py`, is really poor advice.  It's perhaps fine for demonstration purposes as it's understandable what's going on, but should be very clearly recommended against for production purposes.  The docs should instead recommend using the correct `build_ext` command.  I don't think this should be done automatically--I think the onus is on the author of software using Cython to write their `setup.py` correctly.

setuptools does have a thing where if Cython is detected it replaces its build_ext command with the one from Cython automatically.  But that's had some problems lately and I'm not up to date on where that stands now.

**Comment by McSinyx** [comment]
> `setuptools` does have a thing where if Cython is detected it replaces its `build_ext` command with the one from Cython automatically

Current implementation put cythonization on hold untill `build_ext` is called, thus the original issue no longer exists.  I think this can be closed now.
